File: 2016  Plan/5月/PictureInPictureDemo/PictureInPictureDemo/ViewController.m
//
//  ViewController.m
//  PictureInPictureDemo
//
//  Created by Qinting on 16/6/16.
//  Copyright © 2016年 Qinting. All rights reserved.
//

#import "ViewController.h"
#import "PipVideoView.h"
#import <AVKit/AVKit.h>
#import <AVFoundation/AVFoundation.h>

#define currentItemStatus @"currentItem.status"

@interface ViewController ()   // <AVPictureInPictureControllerDelegate>

@property (strong, nonatomic) IBOutlet PipVideoView *pipVideoView;

@property (nonatomic,strong) AVPlayer *player;

@property (nonatomic,strong) AVPictureInPictureController *pipController;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    NSURL *path = [[NSBundle mainBundle] URLForResource:@"video" withExtension:@".mov"];
    AVAsset *asset = [AVAsset assetWithURL:path];
    AVPlayerItem *item = [[AVPlayerItem alloc]initWithAsset:asset];
    [self.player replaceCurrentItemWithPlayerItem:item];
     self.pipVideoView.pipAVPlayer = self.player;
    [self.player play];
}

- (AVPlayer *)player {
	if(_player == nil) {
		_player = [[AVPlayer alloc] init];
        [_player addObserver:self forKeyPath:currentItemStatus options:NSKeyValueObservingOptionNew context:nil];
	}
	return _player;
}

-(void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary<NSString *,id> *)change context:(void *)context
{
        if([keyPath  isEqualToString: currentItemStatus])
        {
              NSNumber *oldNum = change[NSKeyValueChangeOldKey];
            NSNumber *num =  [change valueForKey:NSKeyValueChangeNewKey];
            NSLog(@"num: %@---%@",num,oldNum);
            
            if (num.integerValue != AVPlayerItemStatusReadyToPlay) {
                return;
            }
        }

    self.pipController = [[AVPictureInPictureController alloc]initWithPlayerLayer:self.pipVideoView. avPlayerLayer];
//    self.pipController.delegate = self;
}

@end

File: 2016  Plan/5月/PictureInPicture/PipDemo/ViewController.swift
//
//  ViewController.swift
//  PipDemo
//
//  Created by Qinting on 16/6/5.
//  Copyright © 2016年 Qinting. All rights reserved.
//

import UIKit
import AVKit
import AVFoundation

var currentItemStatus = "currentItem.status"
var mcontext = 0
class ViewController: UIViewController,AVPictureInPictureControllerDelegate {

    @IBOutlet var pipView: PipVideoPlay!
    
    lazy var player:AVPlayer = {
     let p = AVPlayer()
        p.addObserver(self, forKeyPath: currentItemStatus, options: .New, context: &mcontext)
        return p
    }()
    
    override func viewDidLoad() {
        let u  = NSBundle.mainBundle().URLForResource("video", withExtension: ".mov")
        let asset = AVAsset(URL: u!)
        let item = AVPlayerItem(asset: asset)
        player.replaceCurrentItemWithPlayerItem(item)
        pipView.pipPlayerPlay = player
        player.play()
    }

    var pipController : AVPictureInPictureController?
    
    override func observeValueForKeyPath(keyPath: String?, ofObject object: AnyObject?, change: [String : AnyObject]?, context: UnsafeMutablePointer<Void>) {
        if keyPath == currentItemStatus {
            let statusInt = change?[NSKeyValueChangeNewKey] as? NSNumber
            let status = AVPlayerItemStatus(rawValue: statusInt!.integerValue)
            
            if status !=  .ReadyToPlay {
                return
            }
            pipController = AVPictureInPictureController(playerLayer: pipView.pipLayer)
            pipController?.delegate = self;
        }
    }
    
    func pictureInPictureControllerWillStartPictureInPicture(pictureInPictureController: AVPictureInPictureController){
    print("Will Start")
    }
    
    func pictureInPictureControllerDidStartPictureInPicture(pictureInPictureController: AVPictureInPictureController){
     print("Did Start")
    }

    func pictureInPictureController(pictureInPictureController: AVPictureInPictureController, failedToStartPictureInPictureWithError error: NSError){
     print("failedToStart:%@",error)
    }

}

